plenoptic/simulate/models/naive.py

```diff
@@ -159,56 +184,87 @@ class CenterSurround(nn.Module):
+            center_std = torch.ones(out_channels) * center_std
+        if isinstance(surround_std, float) or surround_std.shape == torch.Size([]):
+            surround_std = torch.ones(out_channels) * surround_std
+        assert len(center_std) == out_channels and len(surround_std) == out_channels, "stds must correspond to each out_channel"
         assert width_ratio_limit > 1.0, "stdev of surround must be greater than center"
         assert amplitude_ratio >= 1.0, "ratio of amplitudes must at least be 1."
 
         self.on_center = on_center
+
         self.kernel_size = kernel_size
         self.width_ratio_limit = width_ratio_limit
-        self.register_buffer("amplitude_ratio", torch.tensor(amplitude_ratio))
+        self.register_buffer("amplitude_ratio", torch.as_tensor(amplitude_ratio))
 
-        self.center_std = nn.Parameter(torch.tensor(center_std))
-        self.surround_std = nn.Parameter(torch.tensor(surround_std))
+        self.center_std = nn.Parameter(torch.ones(out_channels) * center_std)
+        self.surround_std = nn.Parameter(torch.ones(out_channels) * surround_std)
 
+        self.out_channels = out_channels
         self.pad_mode = pad_mode
 
+        self.cache_filt = cache_filt
+        self._filt = None
+
     @property
     def filt(self) -> Tensor:
         """Creates an on center/off surround, or off center/on surround conv filter"""
-        filt_center = circular_gaussian2d(self.kernel_size, self.center_std)
-        filt_surround = circular_gaussian2d(self.kernel_size, self.surround_std)
-        on_amp = self.amplitude_ratio
+        if self._filt is not None:  # use cached filt
+            return self._filt
+        else:  # generate new filt and optionally cache
+            filt_center = circular_gaussian2d(self.kernel_size, self.center_std, self.out_channels)
+            filt_surround = circular_gaussian2d(self.kernel_size, self.surround_std, self.out_channels)
+            on_amp = self.amplitude_ratio
+
+            # sign is + or - depending on center is on or off
+            sign = torch.as_tensor([1. if x else -1. for x in self.on_center])
+            sign = sign.view(self.out_channels, 1, 1, 1)
+            filt = on_amp * (sign * (filt_center - filt_surround))
 
-        if self.on_center:  # on center, off surround
-            filt = on_amp * filt_center - filt_surround  # on center, off surround
-        else:  # off center, on surround
-            filt = on_amp * filt_surround - filt_center
+            filt = filt / filt.sum()
 
-        filt = filt / filt.sum()
+            if self.cache_filt:
+                self._filt = filt
 
         return filt
 
-    def _clamp_surround_std(self) -> Tensor:
+    def _clamp_surround_std(self):
         """Clamps surround standard deviation to ratio_limit times center_std"""
-        return self.surround_std.clamp(
-            min=self.width_ratio_limit * float(self.center_std), max=None
-        )
+        # TODO
```

Comment:
confused about this? if it's just returning then why the rest of the code? that should be commented out as well if it's not being used or moved to another PR.

Reply:
Whoops nice catch. Fixed now and is no longer dead code.